### PR TITLE
Find loops via SCCs

### PR DESCRIPTION
**Pull request by Kha**
The only (small) generated assembler difference is that Bad nodes may now induce critical edges elimination. But I guess handling/not emitting dead code is a general problem still to be tackled.

### CODE REVIEW

File: src/main/scala/mjis/opt/FirmExtensions.scala [hunk 1]
```diff
@@ -27,17 +26,10 @@ object FirmExtensions {
...
-        override def visitBlock(block: Block): Unit = {
-          for (pred <- block.getPreds.filter(!_.isInstanceOf[Bad]))
-            m(pred.getBlock.asInstanceOf[Block]) += block
-        }
-      })
-      m.toMap.withDefaultValue(Set.empty)
-    }
+    def getBlockEdges: Map[Block, Seq[Block]] = ListMap(
+      NodeCollector.fromBlockWalk(g.walkBlocks)
+      .map(b => b -> b.getPreds.map(_.block).filter(_ != null).toSeq): _*
```
Comment (jspam) [inline src/main/scala/mjis/opt/FirmExtensions.scala]: Won't `_.block` of a `Bad` node return the start block?

I guess the check in `NodeExt.block` could also be wrong, it should check `node` for `Bad`, not `node.getBlock`.

File: src/main/scala/mjis/Optimizer.scala [hunk 2]
```diff
@@ -18,11 +18,11 @@ class Optimizer(input: Unit) extends Phase[Unit] {
 
   private def removeCriticalEdges(g: Graph): Unit = {
     BackEdges.enable(g)
```
Comment (jspam) [inline src/main/scala/mjis/Optimizer.scala]: Do we even need back edges here?

File: src/main/scala/mjis/util/Digraph.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
+package mjis.util
+
+import scala.collection.Map
+import scala.collection.mutable
+import mjis.util.MapExtensions._
+
+import scala.collection.mutable.ArrayBuffer
+
+object Digraph {
+  def getTopological[V](edges: Map[V, Seq[V]], start: V, nodes: Option[Set[V]] = None, visited: mutable.Set[V] = mutable.Set[V]()): Seq[V] = {
```
Comment (jspam) [inline src/main/scala/mjis/util/Digraph.scala]: I think `getNodesInTopologicalOrder` would be clearer.

File: src/main/scala/mjis/opt/FirmExtensions.scala (hunk 1 above)
Comment (Kha) [inline src/main/scala/mjis/opt/FirmExtensions.scala]: Oh wow, never noticed the bug in `NodeExt`. This seems to fix the Bad critical edges.

**Comment by jspam** [comment]
Apart from the comments, :musical_note: LGTM :musical_note:

File: src/main/scala/mjis/util/Digraph.scala (hunk 3 above)
Comment (Kha) [inline src/main/scala/mjis/util/Digraph.scala]: Compromise? <3

**Pull request merged**